design/mtls-validation-extension.md

```diff
@@ -0,0 +1,46 @@
+# MTLS Validation Extension (MTLS_Ext)
+
+* **Status**: Notes
+* **JIRA**: TBD
+
+
+## Motivation
+
+MTLS validation extension is designed to enable extension of standard SSL Certificate validation already present inside Keycloak (X509) enabling validation of additional fields present in different certificate types (example: QWAC), when MTLS 
+support is
+enabled.
+
+QWAC is a sample of standard certificate extension that is occurring on OPEN BANKING Environment 
+(link to [QWAC spec](https://www.etsi.org/deliver/etsi_ts/119400_119499/119495/01.03.01_60/ts_119495v010301p.pdf)).
+
+Introducing an extra introspection/validation SPI has such main benefits:
+
+* Support validation of additional SSL Certificate fields
+* Support for market specific OpenBanking implementations for example European PSD2 (QWAC) or Australian CDR (normative usage of existing fields)
+* Support custom (proprietary) certificate validation implementations
+
+## Implementation details
+We have to provide new Certificate validation SPI.
+This SPI may be implemented by a separate service (plugin), which deals with custom SSL Certificate processing.
+This plugin should be deployed inside the Keycloak and should handle all custom (proprietary) validation and parsing logic, so that Keycloak only interacts with it via SPI.
```

Comment:
I think that the ExtendedCertificateValidation provider is applied to the client. How do you apply this provider onto the client?
- per keycloak
   One type of ExtendedCertificateValidation provider is applied to all client in all realm in the keycloak.
- per realm
   One type of ExtendedCertificateValidation provider are applied to all client in the realm (eg. realm A) and other one is applied to all client in the other realm (eg. realm B)
- per realm and client
  One type of ExtendedCertificateValidation provider is applied to some client (eg. Client A) in the realm (eg. realm A) and other one is applied to other client (Client B) in the this realm.

I think `per keycloak` is the easiest to implement and it is reasonable because it is not so plausible that keycloak treats several market specific profiles (PSD2, UK OB, CDR) simultaneously (eg. realm A is for PSD2, realm B is for UK OB, realm C is for CDR ...).

Reply:
I actually thought of third option, as you surely might have situations when you need same realm work different for different client example: 1 client works with some backend through back channel using some grant password flow without anything other protection means, 2 client works for some XS2A like open interface with FAPI and mTLS enabled, 3 client is for OnlineBanking OAuth2 browser flow without mTLS support as there can't be any certificate at your own customer disposal.
And yes, you're absolutely right that implementing global support is the easiest option. Here I did not mention possibility of having more than one `extended validator`, but there could be people willing to have multiple additional validators and picking the desired one from UI for respective client

Reply:
Yes, I agree with your proposal, namely the option 3.